src/clients/ethereum.ts

```diff
@@ -34,8 +41,38 @@ export type ContractFilter = {
   filter: string
 };
 
+async function getLogs(provider: JsonRpcProvider, params: any, fromBlock: string, toBlock: string){
+  const events = [];
+  let start = BigNumber.from(fromBlock).toHexString();
+  let end = BigNumber.from(toBlock).toHexString();
+  let readUntil = '0';
+  while (BigNumber.from(toBlock).gt(readUntil)){
+    try {
+      const eventRange = await provider.send('eth_getLogs', [{
+        ...params,
+        fromBlock: start,
+        toBlock: end
+      }]);
+    
+      events.push(...eventRange)
+      readUntil = end;
+      start = end;
+      end = BigNumber.from(toBlock).toHexString();
+    } catch (e: any){
+      const errorString = e.toString() as string;
+      const suggestion = errorString.substring(errorString.indexOf('this block range should work: ['), errorString.indexOf(']\\"}}",'));
+      const suggestedRanges = suggestion.substring(suggestion.indexOf('[') + 1).split(', ')
+      start = suggestedRanges[0];
+      end = suggestedRanges[1];
```

Comment:
not ideal if we're implementing a dependency on alchemy here, but it is nice to have this guiding range. i'd say to keep this but we should add more explicit checks on the error string to confirm it is in fact this exact error, and it contains exactly this string so we can throw a proper error if alchemy isn't being used.

similarily, ensuring that if alchemy for example changes there error message or structure of where they put the guiding block range or makes it json or whatever, we make sure we detect this and throw a proper error too